### PR TITLE
CBG-1971: TestReplicationConcurrentPush flaky test fix

### PR DESCRIPTION
**Pull request by gregns1**
CBG-1971

Test was flaking on waiting for nodes to be assigned. But also did a few runs where that wasn't included and notcied that it also failed on asserting for the number of docs pushed on the second replicator (rep_DEF). After digging into the losg I noticed that in all cases where it failed for this reaon I saw a log line:
`[INF] Replicate: t:TestReplicationConcurrentPush c:sgr-mgr- db:activedb Initializing replication <ud>rep_ABC</ud>`
But no corresponding log line for the second created replications (rep_DEF). When looking into the code I noticed this is called when refreshing the sgreplicate config. After adding new print lines everywhere I noticed on failure scenarios the reason the test wasn't actually initialising the second replication was due to `StartReplications` being called after the first replication was created and inserted into sgreplicate. This function created a new suscription and in this time the second replication config is inserted into sgreplicate. There is a race where this second replication is missed as a new addition as a new channel is created in subscribe replicate. 

## Pre-review checklist
- [x] Removed debug logging (`fmt.Print`, `log.Print`, ...)
- [x] Logging sensitive data? Make sure it's tagged (e.g. `base.UD(docID)`, `base.MD(dbName)`)
- [x] Updated relevant information in the API specifications (such as endpoint descriptions, schemas, ...) in `docs/api`

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [x] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/1648/ (for running particular tests 200 times) 
- [x] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/1650/   (full inegration test run)

### CODE REVIEW

File: db/sg_replicate_cfg.go [hunk 1]
```diff
@@ -1191,6 +1191,10 @@ func (c *SGRCluster) GetReplicationIDsForNode(nodeUUID string) (replicationIDs [
 	return replicationIDs
 }
 
+func (m *sgReplicateManager) GetActiveReplicators() map[string]*ActiveReplicator {
+	return m.activeReplicators
```
Comment (bbrks) [inline db/sg_replicate_cfg.go]: I think we need a lock to read values from this map (`m.activeReplicatorsLock`) and given the function returns the full map, we have no control over the locking.

We'll either need to make a full copy of the map, or given the use case, maybe a method specifically for retrieving `NumActiveReplicators()` which holds the lock whilst calling `len()`

File: rest/replicatortest/replicator_test.go [hunk 2]
```diff
@@ -935,12 +935,14 @@ func TestReplicationConcurrentPush(t *testing.T) {
 
 	activeRT, remoteRT, remoteURLString, teardown := rest.SetupSGRPeers(t)
 	defer teardown()
-	// Create push replications, verify running
+	// Create push replication, wait for assigned replication (this gives time for the first replication to be created and start initializing),
+	// verify running, then create second replication, verify running and verify both replicators are initialized on sgReplicateManager
 	activeRT.CreateReplication("rep_ABC", remoteURLString, db.ActiveReplicatorTypePush, []string{"ABC"}, true, db.ConflictResolverDefault)
-	activeRT.CreateReplication("rep_DEF", remoteURLString, db.ActiveReplicatorTypePush, []string{"DEF"}, true, db.ConflictResolverDefault)
```
Comment (bbrks) [inline rest/replicatortest/replicator_test.go]: Is there actually a bug here where we're not able to create two replications at the same time?

I'd expect the CAS retry in `sgReplicateManager.updateCluster` ensures that both could be created, even if concurrently, as it's something the REST API could potentially allow to happen in the real world.

**Review by bbrks — CHANGES_REQUESTED** [review]
I'm not convinced that we're not just avoiding a real-world race condition here that we'd like to handle better (with CAS retry/update callback)

File: rest/replicatortest/replicator_test.go (hunk 2 above)
Comment (gregns1) [reply rest/replicatortest/replicator_test.go]: So my analysis of the problem (I could be wrong) was that we can create two at the same time, but the changes to sg replicate config come through on a channel `cfgEvents` and we have a race where sometimes where only the first replication is picked up and takes time to initialise. Then by the time the second replication is processed as a cfg change and goes through the refresh config/initialise replication process the test is already asserting against the stats for that replication. I think in real world for customers the replication will initialise eventually. Also think that this test isn't really a problem when running against server I think its more flaking for walrus and I put that down to the added latency when communicating with server helping in the context of the test?

File: db/sg_replicate_cfg.go (hunk 1 above)
Comment (gregns1) [reply db/sg_replicate_cfg.go]: Good point will change this!

File: rest/replicatortest/replicator_test.go (hunk 2 above)
Comment (bbrks) [reply rest/replicatortest/replicator_test.go]: The thing that makes me uncertain of that is that we don't see any logging associated with `rep_DEF` in the entire test (which runs for ~20 seconds)

https://mobile.jenkins.couchbase.com/job/sgw-unix-build/26219/consoleFull 

It's almost like DEF never even gets created (or at least assigned to a node)

**Review by bbrks — APPROVED** [review]
I'm OK with this fix after in-person discussion.

**Pull request merged**